_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+---
+layout: blog-detail
+post-type: blog
+by: Meriam Lachkar, Vincenzo Bazzucchi, Scala Center
+title: Explicit term inference with Scala 3
+---
+
+# Explicit term inference with Scala 3
+
+One of the most striking changes for developers adopting Scala 3 is the introduction
+of new syntaxes to replace the implicit mechanism used in previous Scala versions.
+
+The motivation behind the new syntax is that the same `implicit` keyword was
+used for different purposes and patterns and thus it became a way to express *how* to implement
+patterns. This means that when encountering this ambiguous incantation, users need to decipher what
+the intent of the developer was: is this a conversion ? Does this avoid parameter repetition ?
+Is this an extension of a type ? Is this a typeclass ? How do I import this ?
+
+Seeing how pervasive implicits became in libraries and projects, Scala 3 aims at reducing confusion and cognitive load by using new keywords that convey the intent of the developer.
+
+This post briefly introduces the new syntaxes and semantics available to Scala 3 programmers by analysing the
+most common use cases and patterns: extension methods, implicit parameters, implicit conversions and typeclasses.
+
+## Compatibility disclaimer
+
+While we believe that the new syntax represents an improvement, it is very important to highlight
+that older code using `implicit` is still perfectly valid for the Scala 3.0 compiler, even if it will be deprecated in future releases.
+You do not need to port your codebase right away, it can be an incremental and gradual process.
+
+## Extension methods
+
+When you do not have control over a type but you need to extend its capabilities with a new method,
+Scala 3 allows you to define an *extension method*
+
+Assume that you are working with `List[Try[String]]` and that you often need to retrieve
+the elements for which the computation succeeded.
+
+Then you can extend this type to have a `collectSucceded` method:
+
+```scala
+// ListTryOps.scala
+
+import scala.util.{Try, Success}
+
+extension (ls: List[Try[String]]) def collectSucceeded: List[String] =
+  ls.collect { case Success(x) => x }
+```
+To remember the syntax, remark how `collectSucceeded` follows the object on which it will be available: `ls.collectSucceeded`.
+Note that extension methods can have type parameters as well:
```

Comment:
```suggestion
Extension methods can have type parameters as well:
```
Usually, we wouldn't use "note" to introduce something new. We only use that if it was deductible from what was said before, and we're only bringing attention to it.